lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js

```diff
@@ -117,237 +475,86 @@ module.exports = {
+          if (hasResponseTemplate(event)) {
+            integrationResponseTemplate = event.http.response.template;
           }
 
-          if (corsEnabled) {
-            const corsMethodResponseParameter = {
-              'method.response.header.Access-Control-Allow-Origin':
-                'method.response.header.Access-Control-Allow-Origin',
-            };
+          // 4. Has CORS enabled?
+          if (hasCors(event)) {
+            corsPreflight[path] = generateCorsPreflightConfig(event.http.cors,
+              corsPreflight[path], method);
 
-            const corsIntegrationResponseParameter = {
-              'method.response.header.Access-Control-Allow-Origin':
-              `'${cors.origins.join('\',\'')}'`,
+            const corsHeader = {
+              'Access-Control-Allow-Origin':
+              `'${corsPreflight[path].origins.join('\',\'')}'`,
             };
 
-            _.merge(methodResponses[0].ResponseParameters, corsMethodResponseParameter);
-            _.merge(integrationResponses[0].ResponseParameters, corsIntegrationResponseParameter);
+            _.merge(methodResponseHeaders, corsHeader);
+            _.merge(integrationResponseHeaders, corsHeader);
+          }
+
+          // Sort out response headers
+          if (hasResponseHeaders(event)) {
+            _.merge(methodResponseHeaders, event.http.response.headers);
+            _.merge(integrationResponseHeaders, event.http.response.headers);
           }
 
-          // add default status codes
-          methodResponses.push(
-            { StatusCode: 400 },
-            { StatusCode: 401 },
-            { StatusCode: 403 },
-            { StatusCode: 404 },
-            { StatusCode: 422 },
-            { StatusCode: 500 },
-            { StatusCode: 502 },
-            { StatusCode: 504 }
-          );
-
-          integrationResponses.push(
-            { StatusCode: 400, SelectionPattern: '.*\\[400\\].*' },
-            { StatusCode: 401, SelectionPattern: '.*\\[401\\].*' },
-            { StatusCode: 403, SelectionPattern: '.*\\[403\\].*' },
-            { StatusCode: 404, SelectionPattern: '.*\\[404\\].*' },
-            { StatusCode: 422, SelectionPattern: '.*\\[422\\].*' },
-            { StatusCode: 500,
-              SelectionPattern:
-                '.*(Process\\s?exited\\s?before\\s?completing\\s?request|\\[500\\]).*' },
-            { StatusCode: 502, SelectionPattern: '.*\\[502\\].*' },
-            { StatusCode: 504, SelectionPattern: '.*\\[504\\].*' }
-          );
+          // Sort out response config
+          const responseConfig = {
+            methodResponseHeaders,
+            integrationResponseHeaders,
+            integrationResponseTemplate,
+          };
 
-          const normalizedFunctionName = functionName[0].toUpperCase()
-            + functionName.substr(1);
+          // Merge in any custom response config
+          if (event.http.response && event.http.response.statusCodes) {
```

Comment:
agreed, maybe it's best to use some sort of a merge functionality rather than if/statements.